### ISSUE TITLE
Molstar plugin fails the build when imported in a project with isolatedModules set to true in tsconfig

### ISSUE THREAD

**Issue by azangru** [molstar/molstar#361]
When I tried to use the Molstar plugin following the [docs](https://molstar.org/docs/plugin/) (the "PluginContext without built-in React UI" option), I ran up against the following error during the build:

![image](https://user-images.githubusercontent.com/6834224/151825682-6878e64a-18ae-4506-896b-e61ac9e054bc.png)

This, as it turned out, is due to the fact that Molstar uses enums as consts in file exports, and this language feature is not supported by typescript when the compiler is configured to use the `isolatedModules` option.

[Repro case](https://github.com/azangru/molstar-experiments/tree/1c1d92d7e73eda2e5db9d9466f2ff97f9381291c) — a newly scaffolded project based on Vite, which puts `isolatedModules` in the tsconfig by default.

While I understand the attractiveness of const enums — they are compiled away by typescript resulting in a slightly cleaner and smaller javascript file — won't the fact that Molstar can be used as a plugin by third-party projects with different typescript settings serve as a sufficient reason for avoiding this language feature (e.g. by using regular enums)?

**References:**

- [Relevant section in typescript docs](https://www.typescriptlang.org/tsconfig#isolatedModules) about the incompatibility of const enums with the `isolatedModules` setting in tsconfig 
- A [discussion](https://stackoverflow.com/a/62175702/3925302) on Stack Overflow urging library authors not to export const enums
- A [blog post](https://ncjamieson.com/dont-export-const-enums) by an RxJS core team member making the same case of avoiding const enum exports when building a library

**Comment by arose** [comment]
Yeah, we did consider that before but it wasn't much of a problem for us. I think for most `const enum`s we can just use `enum`. It is not so much that we want a smaller js file but we enums in some tight-ish loops. We would need to do some performance test.

**Comment by arose** [comment]
@azangru would you be willing to do a PR that uses `enum` instead of `const enum` throughout? I can do some performance test using it then. I don't know if there are other issues apart from `const enum`s.

@dsehnal any objections to enabling isolatedModules? You did something related to them before, didn't you.

**Comment by dsehnal** [comment]
I've ran into the issue before too, but I just worked around it locally. 

I think we could just find the instances where we use the const enum in the loops and optimize those. 

There might also be more problems with the isolatedModules related to how we use namespaces so that it will require a bit more refactoring.

**Comment by azangru** [comment]
> @azangru would you be willing to do a PR that uses enum instead of const enum throughout?

I can try 🙂

**Comment by azangru** [comment]
Right. A simple replacement of `export const enum` with `export enum` wouldn't work. When running the tests, I got errors like the one below, where jest was complaining that it couldn't read the enum off the namespace:

<img width="618" alt="image" src="https://user-images.githubusercontent.com/6834224/153735081-ffad5f8f-43ce-4be3-a8d7-7732109fec0c.png">

I guess the problem is that namespaces, along with other typescript code, are removed during the compilation and do not exist in the javascript runtime; but you are referencing enums as variables that must be accessible at runtime. It all works out if typescript compiler transpiles the source code in its entirety; but if it does the compilation of isolated modules, it cannot access the enums on the namespaces that have been compiled away.

The proper solution would involve moving the enums out of the namespaces, such that they are referenced directly rather than through the namespace name (e.g. `Category.Residue` rather than `ColorTheme.Category.Residue`). Don't know how happy you will be with this refactoring. (Ideally, one wouldn't use namespaces at all of course, given how typescript practitioners generally discourage the use of this old typescript feature; but I see that they are heavily used throughout the Molstar codebase.)

**Comment by dsehnal** [comment]
If we do this refactoring, we would need to release 4.0, which isn't necessarily a bad thing. I can definitely see the value of supporting the ``isolatedModules`` option. Together with this, having a "library build" of Mol* (which could be imported as a ``<script>`` and used from JS without having to use webpack) could bring a lot of value to users.

If you are willing to help us with this @azangru you are more than welcome. Although before we start with such big refactoring, we should probably have at least some basic roadmap of how all this would look like.

Thoughts @arose  @azangru ?

**Comment by azangru** [comment]
Yes, I thought you might need a major version bump, because moving enums outside of the namespaces would be a breaking change for end users. Having a library build of Molstar sounds great! Happy to help.

**Comment by azangru** [comment]
Ah, the problem is even larger than I thought. There are plenty of places in the Molstar codebase where variables are declared and exported inside namespaces ([example](https://github.com/molstar/molstar/blob/b53debcfef324066f71ba0f60a8b3f45ef6f2a31/src/mol-model-formats/structure/property/symmetry.ts#L19-L24)). But, namespaces are erased during compilation, leaving the variables declared inside them inaccessible to the javascript code.

**Comment by azangru** [comment]
@dsehnal @arose I've been wondering if you've had a chance to think about a roadmap that would make molstar a more portable library, or at least will become a stepping stone towards that goal?

On my part, I can see several changes that should be mostly mechanical, but can result in breaking changes of the api:

### Namespaces
Just remove them. They are a very old, and unfortunate, feature of typescript; a relict, if I remember correctly, of pre-ES6 days, before javascript modules got standardised. Unless I am mistaken, they offer no benefits, and these days are treated as one of the "bad parts" of typescript.

Such removal can sometimes be done without breaking changes. For example, in [src/mol-util/value-cell.ts](https://github.com/molstar/molstar/blob/ab4a24d8aba81cb2b90d70e161c83c62c31814f5/src/mol-util/value-cell.ts), it should probably be sufficient (although I haven't checked) to change the namespaces `ValueCell`, `ValueBox`, and `ValueRef` into plain objects, such that this code:

```ts
interface ValueRef<T extends any = unknown> { ref: T }

namespace ValueRef {
    export function create<T>(ref: T): ValueRef<T> { return { ref }; }
    export function set<T>(ref: ValueRef<T>, value: T) { ref.ref = value; return ref; }
}

export { ValueRef }
```

changes into this:

```ts
type ValueRef<T> = {
  ref: T;
};

const createValueRef = <T>(ref: T): ValueRef<T> => ({ ref });
const setValueRef = <T>(ref: ValueRef<T>, value: T): ValueRef<T> => {
  ref.ref = value;
  return ref;
};


const ValueRefMethods = {
  create: createValueRef,
  set: setValueRef
};

export { ValueRefMethods as ValueRef };
```

A somewhat more confusing example in [src/mol-canvas3d/camera/stereo.ts](https://github.com/molstar/molstar/blob/3b1513adc0048dc4879f1d70874b3e56aaffd10e/src/mol-canvas3d/camera/stereo.ts#L22-L56), where the function `is` is defined inside the StereoCamera namespace can probably be rewritten by making `is` into a static method on the `StereoCamera` class. Although I am not sure whether this function is even used at all.

### Enums
There are currently cases in which enums are declared inside namespaces

```ts
namespace NS {

  export const enum Foo {
    FOO,
    BAR,
    BAZ
  }

}

export { NS }
```

I can only think of either moving the enums out of the namespaces; or changing them into constants, like so:

```ts
export const IntDataType {
    Int8: 1,
    Int16: 2,
    Int32: 3,
    Uint8: 4,
    Uint16: 5,
    Uint32: 6,
}

const Encoding = {
    IntDataType
}

export { Encoding }
```

Both of these options will be a breaking change: if an enum is moved out of the namespace, it will have to be referenced without the namespace; whereas if it is changed into a const and kept inside an object simulating the namespace, it will have to be referenced in type declarations as `typeof Encoding.IntDataType[keyof Encoding.IntDataType]`. A utility type can be created to simplify this boilerplate.


### Types declared inside namespaces

Example: [src/mol-io/common/binary-cif/encoding.ts](https://github.com/molstar/molstar/blob/b53debcfef324066f71ba0f60a8b3f45ef6f2a31/src/mol-io/common/binary-cif/encoding.ts#L58-L155). I am not sure how best to approach those.



Any thoughts? Is this something that you would like to pursue?

**Comment by arose** [comment]
I am all for it but don't have time code it myself. I am OK to review PRs. Maybe one way to start is to do a series of PRs, starting with things that can be done without breaking changes. For breaking changes we can target a major version (e.g. 4.0). We should also add eslint rules to make the use namspaces and enums first a warning (for the transition/development phase) and then an error.
